compiler/codegen/ELFObjectFileGenerator.cpp

```diff
@@ -0,0 +1,307 @@
+   trlr->dynsymSection.sh_offset = symbolsStartOffset;
+   trlr->dynsymSection.sh_size = (numELFSymbols() + 1)*sizeof(ELFSymbol);
+   trlr->dynsymSection.sh_link = 5; // dynamic string table index
+   trlr->dynsymSection.sh_info = 1; // index of first non-local symbol: for now all symbols are global
+   trlr->dynsymSection.sh_addralign = 8;
+   trlr->dynsymSection.sh_entsize = sizeof(ELFSymbol);
+
+   trlr->shstrtabSection.sh_name = trlr->shstrtabSectionName - trlr->zeroSectionName;
+   trlr->shstrtabSection.sh_type = SHT_STRTAB;
+   trlr->shstrtabSection.sh_flags = 0;
+   trlr->shstrtabSection.sh_addr = 0;
+   trlr->shstrtabSection.sh_offset = trailerStartOffset + offsetof(ELFTrailer, zeroSectionName);
+   trlr->shstrtabSection.sh_size = sizeof(trlr->zeroSectionName) +
+                                   sizeof(trlr->shstrtabSectionName) +
+                                   sizeof(trlr->textSectionName) +
+                                   sizeof(trlr->relaSectionName) +
+                                   sizeof(trlr->dynsymSectionName) +
+                                   sizeof(trlr->dynstrSectionName);
+   trlr->shstrtabSection.sh_link = 0;
+   trlr->shstrtabSection.sh_info = 0;
+   trlr->shstrtabSection.sh_addralign = 1;
+   trlr->shstrtabSection.sh_entsize = 0;
+
+   trlr->dynstrSection.sh_name = trlr->dynstrSectionName - trlr->zeroSectionName;
+   trlr->dynstrSection.sh_type = SHT_STRTAB;
+   trlr->dynstrSection.sh_flags = 0;
+   trlr->dynstrSection.sh_addr = 0;
+   trlr->dynstrSection.sh_offset = symbolNamesStartOffset;
+   trlr->dynstrSection.sh_size = _totalELFSymbolNamesLength;
+   trlr->dynstrSection.sh_link = 0;
+   trlr->dynstrSection.sh_info = 0;
+   trlr->dynstrSection.sh_addralign = 1;
+   trlr->dynstrSection.sh_entsize = 0;
+
+   trlr->zeroSectionName[0] = 0;
+   strcpy(trlr->shstrtabSectionName, ".shstrtab");
+   strcpy(trlr->textSectionName, ".text");
+   strcpy(trlr->relaSectionName, ".rela.text");
+   strcpy(trlr->dynsymSectionName, ".symtab");
+   strcpy(trlr->dynstrSectionName, ".strtab");
+
+   // now walk list of compiled code symbols building up the symbol names and filling in array of ELFSymbol structures
+   ELFSymbol *elfSymbols = trlr->symbols + 0;
+   char *elfSymbolNames = pointer_cast<char *>(elfSymbols + (numELFSymbols()+1));
+
+   // first symbol is UNDEF symbol: all zeros, even name is zero-terminated empty string
+   elfSymbolNames[0] = 0;
+   elfSymbols[0].st_name = 0;
+   elfSymbols[0].st_info = ELF_ST_INFO(0,STT_NOTYPE);
+   elfSymbols[0].st_other = 0;
+   elfSymbols[0].st_shndx = 0;
+   elfSymbols[0].st_value = 0;
+   elfSymbols[0].st_size = 0;
+
+   ELFSymbol *elfSym = elfSymbols + 1;
+   char *names = elfSymbolNames + 1;
+   for (auto it = _symbols.begin(); it != _symbols.end(); ++it)
+      {
+      fprintf(stderr, "Writing elf symbol %d, name(%d) = %s\n", (elfSym - elfSymbols), it->_nameLength, it->_name);
```

Comment:
This comment applies to all the `fprintf(stderr, ...)`'s used in this PR.  While useful, I think they should be guarded with some kind of trace option to control the amount of output.